src/foam/nanos/crunch/CapabilityPreconditionDAO.js

```diff
@@ -0,0 +1,58 @@
+/**
+ * @license
+ * Copyright 2020 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.crunch',
+  name: 'CapabilityPreconditionDAO',
+  extends: 'foam.dao.ObjectMutationProxyDAO',
+  flags: ['java'],
+
+  javaImports: [
+    'foam.core.X',
+    'foam.dao.DAO',
+    'foam.nanos.crunch.CrunchService',
+    'static foam.mlang.MLang.*'
+  ],
+
+  documentation: `
+    Replaces the visibilityPredicate by checking if preconditions are satisfied.
+    If preconditions are not satisfied, visibilityPredicate becomes FALSE,
+    otherwise it is not modified.
+  `,
+
+  axioms: [
+    {
+      name: 'javaExtras',
```

Comment:
does this really need to be a javeExtra? Isn't it the same thing as just defining another method.